src/main/java/org/sonarlint/intellij/core/ServerIssueUpdater.java

```diff
@@ -0,0 +1,130 @@
+import java.util.stream.StreamSupport;
+import org.sonarlint.intellij.config.project.SonarLintProjectSettings;
+import org.sonarlint.intellij.issue.IssuePointer;
+import org.sonarlint.intellij.issue.IssueStore;
+import org.sonarlint.intellij.issue.ServerIssuePointer;
+import org.sonarlint.intellij.ui.SonarLintConsole;
+import org.sonarlint.intellij.util.SonarLintUtils;
+import org.sonarsource.sonarlint.core.client.api.connected.ConnectedSonarLintEngine;
+import org.sonarsource.sonarlint.core.client.api.connected.ServerIssue;
+import org.sonarsource.sonarlint.core.client.api.exceptions.SonarLintWrappedException;
+
+public class ServerIssueUpdater extends AbstractProjectComponent {
+
+  private static final int THREADS_NUM = 5;
+
+  private ExecutorService executorService;
+
+  private final IssueStore store;
+
+  public ServerIssueUpdater(Project project, IssueStore store) {
+    super(project);
+    this.store = store;
+  }
+
+  public void trackServerIssues(Set<VirtualFile> virtualFiles) {
+    virtualFiles.forEach(this::trackServerIssues);
+  }
+
+  private void trackServerIssues(VirtualFile virtualFile) {
+    SonarLintProjectSettings projectSettings = SonarLintUtils.get(myProject, SonarLintProjectSettings.class);
+    if (!projectSettings.isBindingEnabled()) {
+      // not in connected mode
+      return;
+    }
+
+    String serverId = projectSettings.getServerId();
+    String moduleKey = projectSettings.getProjectKey();
+    if (serverId == null || moduleKey == null) {
+      // not bound to SQ project
+      return;
+    }
+
+    ConnectedSonarLintEngine engine = SonarLintUtils.get(SonarLintServerManager.class).getConnectedEngine(serverId);
+
+    String relativePath = getRelativePath(virtualFile);
+
+    fetchAndMatchServerIssues(virtualFile, engine, moduleKey, relativePath);
+  }
+
+  private void fetchAndMatchServerIssues(VirtualFile virtualFile, ConnectedSonarLintEngine engine, String moduleKey, String relativePath) {
+    // TODO make it possible to cancel
+    this.executorService.submit(() -> {
+      Iterator<ServerIssue> serverIssues = fetchServerIssues(engine, moduleKey, relativePath);
+
+      Collection<IssuePointer> serverIssuePointers = toStream(serverIssues).map(ServerIssuePointer::new).collect(Collectors.toList());
+
+      if (!serverIssuePointers.isEmpty()) {
+        store.storeServerIssues(virtualFile, serverIssuePointers);
+      }
+    });
+  }
+
+  private <T> Stream<T> toStream(Iterator<T> iterator) {
+    Iterable<T> iterable = () -> iterator;
+    return StreamSupport.stream(iterable.spliterator(), false);
+  }
+
+  private Iterator<ServerIssue> fetchServerIssues(ConnectedSonarLintEngine engine, String moduleKey, String relativePath) {
+    try {
+      return engine.downloadServerIssues(moduleKey, relativePath);
+    } catch (SonarLintWrappedException e) {
+      // download failed, fall back to local storage, if exists
+      return engine.getServerIssues(moduleKey, relativePath);
+    } catch (Throwable t) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)